### PR TITLE
Finish trace when newclientsteram returns error

### LINKED ISSUE #812: Stream fail-fast leaks tracers
If tracing is enabled and we hit any of the early returns in [this block](https://github.com/grpc/grpc-go/blob/35896af9ad39c1fb1b1cd925fe3621be361e3d81/stream.go#L146-L179), the trace is leaked because we haven't started the goroutine that will call `Finish` on it. Each of the error returns between the creation of the tracer and the goroutine needs to call `finish`.

Discovered in cockroachdb/cockroach#8340

### PR DESCRIPTION
**Pull request by menghanl**
fixes #812

### CODE REVIEW

File: stream.go [hunk 1]
```diff
@@ -140,6 +139,13 @@ func NewClientStream(ctx context.Context, desc *StreamDesc, cc *ClientConn, meth
 		cs.trInfo.tr.LazyLog(&cs.trInfo.firstLine, false)
 		ctx = trace.NewContext(ctx, cs.trInfo.tr)
 	}
+	defer func() {
+		if err != nil {
+			// Need to call cs.finish() if error is returned.
+			// Because cs will not be returned to caller.
+			cs.finish(err)
```
Comment (tamird) [inline stream.go]: `cs` is not even used until way down below after the for loop. Why does it need to be allocated way up here?

Comment (menghanl) [inline stream.go]: Done. Trace needs to start before the for loop, so it's split out.

**Comment by tamird** [comment]
LGTM

**Comment by tamird** [comment]
@menghanl @iamqizhao can you get this in please? It's currently preventing us from debugging an issue in CockroachDB.

**Pull request merged**